modules/git/repo.go

```diff
@@ -425,14 +425,20 @@ func (repo *Repository) CreateBundle(ctx context.Context, commit string, out io.
 	}
 	defer os.RemoveAll(tmp)
 
-	tmpFile := filepath.Join(tmp, "bundle")
-	args := []string{
-		"bundle",
-		"create",
-		tmpFile,
-		commit,
+	env := append(os.Environ(), "GIT_OBJECT_DIRECTORY="+filepath.Join(repo.Path, "objects"))
+	_, err = NewCommandContext(ctx, "init", "--bare").RunInDirWithEnv(tmp, env)
+	if err != nil {
+		return err
 	}
-	_, err = NewCommandContext(ctx, args...).RunInDir(repo.Path)
+
+	branchName := "main"
+	_, err = NewCommandContext(ctx, "branch", branchName, commit).RunInDirWithEnv(tmp, env)
+	if err != nil {
+		return err
+	}
+
+	tmpFile := filepath.Join(tmp, "bundle")
+	_, err = NewCommandContext(ctx, "bundle", "create", tmpFile, branchName).RunInDirWithEnv(tmp, env)
```

Comment:
Hardcoding `main` here is confusing. If I have `main` and `dev` branches and download a bundle of `dev`, it'll show the branch name `main` with the contents of `dev`.

Ideally the bundle would contain two references, `HEAD` and the name of the branch downloaded if we know that (pointing to the same commit). The `HEAD` is a convenience so people cloning from the bundle don't have to specify a branch.

That would require changing the repo archiver machinery to pass a desired branch name into this function, so for an initial fix I think we should just make a bundle with a `HEAD`.

```suggestion
	_, err = NewCommandContext(ctx, "reset", "--soft", commit).RunInDirWithEnv(tmp, env)
	if err != nil {
		return err
	}

	tmpFile := filepath.Join(tmp, "bundle")
	_, err = NewCommandContext(ctx, "bundle", "create", tmpFile, "HEAD").RunInDirWithEnv(tmp, env)
```

Reply:
Clone from such a bundle will lead to repo with detached HEAD. Probably better set "bundle" reference

> Hardcoding main here is confusing. If I have main and dev branches and download a bundle of dev, it'll show the branch name main with the contents of dev
yes, reponame-dev.bundle with main branch inside it.

> That would require changing the repo archiver machinery to pass a desired branch name into this function, so for an initial fix I think we should just make a bundle with a HEAD.
It is not a problem to send the name of the branch, the problem if that the different branches include the same commit, but the commit field is unique in the database.

Reply:
I see, okay.

Perhaps in the future it would be possible to generate the ref headers on the fly, while reusing the existing bundle file. [The bundle format](https://git-scm.com/docs/bundle-format) is not too complicated (essentially a few header lines concatenated with a Git packfile), so we may be able to:

1. Read the headers of the existing bundle
2. Use that info to construct a new header with the correct ref and send that to the client
3. Seek to the end of the header section in the bundle and send that to the client

Without impacting download time/storage.

(Of course, I don't mean to suggest that you need to do any of this.)

Reply:
Better to create an issue with this proposal, so it doesn't get lost